syntax = "proto3";
package animation_ml;

service TextToMotionService {
    rpc GenerateMotion (MotionRequest) returns (MotionResponse);
    rpc GetServerStatus (StatusRequest) returns (StatusResponse);
}

message MotionRequest {
    string prompt = 1;
    float duration_seconds = 2;
    int32 target_fps = 3;
    SkeletonType skeleton_type = 4;
    repeated BoneMapping bone_mappings = 5;
}

enum SkeletonType {
    SMPL_22 = 0;
    VRM_HUMANOID = 1;
}

message BoneMapping {
    int32 source_joint_index = 1;
    string target_bone_name = 2;
}

message MotionResponse {
    repeated AnimationCurve curves = 1;
    float generation_time_ms = 2;
    string model_used = 3;
}

message AnimationCurve {
    string bone_name = 1;
    PropertyType property_type = 2;
    repeated CurveKeyframe keyframes = 3;
}

enum PropertyType {
    TRANSLATION_X = 0;
    TRANSLATION_Y = 1;
    TRANSLATION_Z = 2;
    ROTATION_X = 3;
    ROTATION_Y = 4;
    ROTATION_Z = 5;
}

message CurveKeyframe {
    float time = 1;
    float value = 2;
    float tangent_in_dt = 3;
    float tangent_in_dv = 4;
    float tangent_out_dt = 5;
    float tangent_out_dv = 6;
    InterpolationType interpolation = 7;
}

enum InterpolationType {
    LINEAR = 0;
    BEZIER = 1;
}

message StatusRequest {}

message StatusResponse {
    bool ready = 1;
    string active_model = 2;
    int32 gpu_memory_mb = 3;
}
